htlcswitch/link.go

```diff
@@ -1150,6 +1228,132 @@ func (l *channelLink) htlcManager() {
+						code: ErrInternalError,
+					}, "failed persisting script: %v", err)
+					return
+				}
+			}
+
+			// Mark that we've sent the Shutdown message. This is
+			// used for retransmission.
+			err := l.channel.State().MarkShutdownSent()
+			if err != nil {
+				l.fail(LinkFailureError{
+					code: ErrInternalError,
+				}, "failed marking shutdown: %v", err)
+				return
+			}
+
+			// Notify peer.Brontide that we've sent the Shutdown
+			// message so that the coop close state machine can
+			// advance.
+			closeReq := l.localCloseReq
+			if closeReq == nil {
+				// If we are not the initiator of the coop
+				// close, we'll create a ChanClose struct. We
+				// won't listen on the error or update chans.
+				// If there's an error, we'll eventually stop
+				// the link anyways. The Updates chan is
+				// currently unused, but is added to prevent a
+				// future change from hanging if it attempts to
+				// send along the chan.
+				closeType := contractcourt.CloseRegular
+				chanPoint := l.ChannelPoint()
+				deliveryAddr := l.cfg.DeliveryAddr
+				errChan := make(chan error, 1)
+				updatesChan := make(chan interface{}, 2)
+
+				// Set the TargetFeePerKw to ensure that if a
+				// bug occurs and this struct is used, that the
+				// zero fee isn't used instead.
+				feeEst := l.cfg.FeeEstimator
+				feePerKw, err := feeEst.EstimateFeePerKW(6)
+				if err != nil {
+					l.fail(LinkFailureError{
+						code: ErrInternalError,
+					}, "failed fetching fee: %v", err)
+					return
+				}
+
+				closeReq = &ChanClose{
+					CloseType:      closeType,
+					ChanPoint:      chanPoint,
+					DeliveryScript: deliveryAddr,
+					Err:            errChan,
+					Updates:        updatesChan,
+					TargetFeePerKw: feePerKw,
+				}
+			} else {
+				// The script may have changed so set the
+				// DeliveryScript to the link's DeliveryAddr.
+				closeReq.DeliveryScript = l.cfg.DeliveryAddr
+			}
+
+			// This is called before actually sending Shutdown so
+			// that the coop close state machine does not end up
+			// racing with the ClosingSigned from the peer:
+			//
+			// Current flow:
+			// - State machine receives our Shutdown message
+			// - Send Shutdown message to peer
+			// - Peer replies with ClosingSigned
+			//
+			// Racing flow:
+			// - Send shutdown message to peer
+			// - Peer replies with ClosingSigned
+			// - State machine receives our Shutdown message
+			l.cfg.NotifySendingShutdown(closeReq, l.quit)
+
+			shutdownMsg := lnwire.NewShutdown(
+				l.ChanID(), l.cfg.DeliveryAddr,
+			)
+
+			err = l.cfg.Peer.SendMessage(false, shutdownMsg)
+			if err != nil {
+				// We can just let the peer.Brontide object
+				// stop the link.
+				l.log.Errorf("failed sending shutdown to "+
+					"peer: %v", err)
+			}
+
+			l.shutdownSent = true
+		}
+
+		// If we've sent and received Shutdown and the channel is
+		// clean, then we'll notify peer.Brontide that the channel can
+		// be coop closed and stop the link.
```

Comment:
Hmm, ok so then a major change here is that the _link_ is the one that sends the shutdown message since it can better gauge when the coast is _eventually_ clear?

Reply:
TBH, I prefer having more of the logic in the co-op close state machine as it lets us seperate concerns a bit more cleanly.

Reply:
Review on these parts probably belongs in the PR this one depends on.

- `shutdown` has to be sent when there are _no_ pending updates (like a `update_fail_htlc` that hasn't been committed)
- if the chancloser is responsible for sending `shutdown`, then the link must pause sending any update message while things are bubbled down to the chancloser and then bubbled back up to the link.